zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/flows.kt

```diff
@@ -86,3 +91,65 @@ internal class FlowSerializer<T>(
 
   override fun hashCode() = delegateSerializer.hashCode()
 }
+
+/**
+ * This serializes [StateFlow] instances (not directly serializable) by converting them to
+ * [StateFlowZiplineService] instances which can be serialized by [ziplineServiceSerializer] to
+ * pass-by-reference.
+ */
+@PublishedApi
+internal class StateFlowSerializer<T>(
+  private val delegateSerializer: KSerializer<StateFlowZiplineService<T>>,
+) : KSerializer<StateFlow<T>> {
+  override val descriptor = delegateSerializer.descriptor
+
+  override fun serialize(encoder: Encoder, value: StateFlow<T>) {
+    val service = value.toZiplineService()
+    return encoder.encodeSerializableValue(delegateSerializer, service)
+  }
+
+  private fun StateFlow<T>.toZiplineService(): StateFlowZiplineService<T> {
+    return object : StateFlowZiplineService<T> {
+      override suspend fun collect(collector: FlowZiplineCollector<T>) {
+        try {
+          this@toZiplineService.collect(collector::emit)
+        } finally {
+          collector.close()
+        }
+      }
+
+      override val value: T get() = this@toZiplineService.value
+
+      override fun toString() = this@toZiplineService.toString()
+    }
+  }
+
+  override fun deserialize(decoder: Decoder): StateFlow<T> {
+    val service = decoder.decodeSerializableValue(delegateSerializer)
+    return service.toStateFlow()
+  }
+
+  private fun StateFlowZiplineService<T>.toStateFlow(): StateFlow<T> {
+    return object : StateFlow<T> {
+      override suspend fun collect(collector: FlowCollector<T>): Nothing {
+        channelFlow {
```

Comment:
Using the `take` operator (and others) fails due to `AbortFlowException` not being caught [here](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/common/src/flow/operators/Limit.kt#L67). Wrapping it in a `channelFlow` similar to `FlowSerializer` solves the issue, though I'm not fully sure why.